orchagent/errororch.cpp

```diff
@@ -0,0 +1,558 @@
+     bool updateEntry = false;
+
+     /* Extract current return code and operation */
+     for (size_t i = 0; i < values.size(); i++)
+     {
+         if(fvField(values[i]) == "operation")
+         {
+             strOp = fvValue(values[i]);
+         }
+         else if(fvField(values[i]) == "rc")
+         {
+             strRc = fvValue(values[i]);
+         }
+     }
+
+     /* Get ERROR DB table object */
+     std::shared_ptr<Table> table;
+     string errTableName = getErrorTableName(tableName);
+     if(m_TableNameObjMap.find(errTableName) == m_TableNameObjMap.end())
+     {
+         SWSS_LOG_INFO("Failed to find EHF Table object for %s", errTableName.c_str());
+         return;
+     }
+
+     table = m_TableNameObjMap[errTableName];
+     if(table == NULL)
+     {
+         SWSS_LOG_INFO("Invalid EHF Table object found for %s", errTableName.c_str());
+         return;
+     }
+
+     /* Check if the entry with the key is present in ERROR DB */
+     std::vector<FieldValueTuple> ovalues;
+     if(table->get(key, ovalues))
+     {
+         entryFound = true;
+     }
+
+     if(strRc == "SWSS_RC_SUCCESS")
+     {
+         /* Remove the entry if present in error database */
+         if(entryFound == true)
+         {
+             removeEntry = true;
+         }
+     }
+     else
+     {
+         if(strOp == "create")
+         {
+             addEntry = true;
+         }
+         else if(strOp == "remove")
+         {
+             if(entryFound == true)
+             {
+                 removeEntry = true;
+             }
+             else
+             {
+                 addEntry = true;
+             }
+         }
+         else if(strOp == "set")
+         {
+             if(entryFound == true)
+             {
+                 updateEntry = true;
+             }
+             else
+             {
+                 addEntry = true;
+             }
+         }
+     }
+     SWSS_LOG_DEBUG("key %s operation %s RC %s Entry in DB %d add:remove:update %d:%d:%d ",
+             key.c_str(), strOp.c_str(), strRc.c_str(),
+             entryFound, addEntry, removeEntry, updateEntry);
+     if(addEntry == true || updateEntry == true)
+     {
+         /* Create/Update the database entry */
+         table->set(key, values);
+         SWSS_LOG_INFO("Publish %s(ok) to error db", key.c_str());
+     }
+     else if (removeEntry == true)
+     {
+         /* Remove the entry from database */
+         table->del(key);
+         SWSS_LOG_INFO("Removed %s(ok) from error db", key.c_str());
+     }
+
+     return;
+ }
+
+/* The following function is used by other Orch Agents to add the entry into *
+ * error database and optionally send a notification to the applications     */
+void ErrorOrch::addErrorEntry(sai_object_type_t object_type,
+        std::vector<FieldValueTuple> &appValues, uint32_t flags)
+{
+    SWSS_LOG_ENTER();
+
+    std::shared_ptr<Table> table;
+
+    /*
+     * The following steps are performed:
```

Comment:
Same as above, the comments are helpful but putting the different steps into explicit methods would help with readability.

Reply:
Done